File: sultex/events/work_order.py
import frappe
import json
from erpnext.stock.get_item_details import get_conversion_factor


@frappe.whitelist()
def get_uom(item_code):
    return frappe.db.get_value('Item', item_code, 'stock_uom')


@frappe.whitelist()
def get_req_items_details(item_code,bom_no):
    bom = frappe.get_doc("BOM",bom_no)
    for item in bom.items:
        if item.item_code == item_code:
           return item.uom , item.qty

@frappe.whitelist()
def fetch_items_from_sales_order(sales_order):
    sales_order = json.loads(sales_order)
    so = frappe.get_doc('Sales Order', sales_order[0])
    item = so.items[0]
    if frappe.db.exists('Item', {
        'name': item.item_code,
        'is_stock_item': 1,
        'default_bom': ['!=', '']
    }):
        return {'item': item.item_code, 'qty': item.qty}
    else:
        frappe.msgprint(
            'Make sure selected item is a Stock Item and has a Default Bom set')


@frappe.whitelist()
def stop_execution(expected_delivery_date,planned_start_date,name):
    if expected_delivery_date < planned_start_date:
        val=None
        frappe.db.set_value('Work Order', name, 'expected_delivery_date',val)
        frappe.throw('please make sure that expected delivery date is smaller then planned start date')

@frappe.whitelist()
def fetch_items_from_work_order(work_order):
    work_order = json.loads(work_order)
    wo = frappe.get_doc('Work Order', work_order[0])
    if len(wo.required_items) == 1:
        item = wo.required_items[0]
        if frappe.db.exists('Item', {
            'name': item.item_code,
            'is_stock_item': 1,
            'default_bom': ['!=', '']
        }):
            return {'item': item.item_code, 'qty': item.required_qty}
        else:
            frappe.msgprint(
                'Make sure selected item is a Stock Item and has a Default Bom set')
    else:
        frappe.msgprint('Invalid Selection')


@frappe.whitelist()
def sales_order_query(sales_order,work_order):
    data=[]
    result = frappe.db.get_list('Sales Order Item',
                                filters={
                                    'parent': sales_order
                                },
                                fields=['*'],
                                order_by='idx')
    for row in result:
        wo_list=frappe.db.get_list("Work Order",
                                filters={
                                    'name': ['not in',[work_order]],
                                    'docstatus':1,
                                    'expected_delivery_date':row.delivery_date,
                                    'production_item':row.item_code,
                                    'sales_order':sales_order,
                                },
                                fields=['*']
                                )
        total=0
        qty=0
        for itm in wo_list:
            total+=itm.qty
        qty=row.qty-total
        if qty > 0:
            row.update({'qty':qty})
            data.append(row)

    return data


@frappe.whitelist()
def work_order_query():
    result = frappe.db.get_list('Work Order Item',  filters={'item_allocated': 0},
                                fields=['item_code', 'item_name', 'required_qty', 'parent'])
    return result


@frappe.whitelist()
def get_secondary_uom(item_code):
    item = frappe.get_doc("Item", item_code)
    uom = []

    if len(item.uoms) > 1:
        for row in item.uoms:
            cf = get_conversion_factor(item.item_code, row.uom)

            new_row = {
                "uom": row.uom,
                "conversion_factor": cf.get("conversion_factor")
            }

            uom.append(new_row)

    return uom


def validate(doc, method=None):
    if doc.get("from_work_order"):
        frappe.db.set_value('Work Order Item', {'parent': doc.from_work_order, 'item_code': doc.production_item}, 'item_allocated', '1')
        frappe.db.commit()
        
    bom = frappe.get_doc("BOM",doc.bom_no)
    for item in doc.required_items:
        for row in bom.items:
            if item.item_code == row.item_code:
                item.required_qty = row.qty * doc.qty
                item.uom = row.uom
    doc.work_order_secondary_uom = []
    item = frappe.get_doc("Item", doc.production_item)
    for uom in item.uoms:
        um = doc.append("work_order_secondary_uom")
        um.uom = uom.uom
        um.conversion_factor = uom.conversion_factor
        um.qty = uom.conversion_factor * doc.qty
    if doc.tracking_id in frappe.db.get_list("Work Order", filters={"from_work_order":["!=", doc.name]}, fields=["tracking_id"]):
        frappe.throw("Tracking ID must be unique!")

def submit(doc, method=None):
    if not doc.dynamic_work_order:
        items = get_all_items(doc)
        wk=[]
        for row in items:
            if row['bom_no']:
                project=secondary_qty=secondary_uom=wip_warehouse=fg_warehouse=scrap_warehouse=''
                if doc.project:
                    project = doc.project
                if doc.secondary_qty:
                    secondary_qty = doc.secondary_qty
                if doc.secondary_uom:
                    secondary_uom = doc.secondary_uom
                if doc.wip_warehouse:
                    wip_warehouse = doc.wip_warehouse
                if doc.fg_warehouse:
                    fg_warehouse = doc.fg_warehouse
                if doc.scrap_warehouse:
                    scrap_warehouse = doc.scrap_warehouse
                work_order_doc = frappe.get_doc({
                    'doctype': 'Work Order',
                    'naming_series':doc.naming_series,
                    'status': 'Draft',
                    'tracking_id':doc.tracking_id,
                    'company': doc.company,
                    'production_item':row['item_code'],
                    'qty':row['qty']*doc.qty,
                    'custom_stock_uom':row['stock_uom'],
                    'bom_no':row['bom_no'],
                    'description':row['description'],
                    'dynamic_work_order':doc.name,
                    'project':project,
                    'secondary_qty':secondary_qty,
                    'expected_delivery_date':doc.expected_delivery_date,
                    'secondary_uom':secondary_uom,
                    'wip_warehouse':wip_warehouse,
                    'fg_warehouse':fg_warehouse,
                    'scrap_warehouse':scrap_warehouse,
                    'planned_start_date':doc.planned_start_date,
                    'reference_work_order':doc.reference_work_order,
                })
                work_order_doc.save()
                wk.append(work_order_doc)



def get_bom(item):
    bom_exists=frappe.db.exists('BOM', {'item':item,'is_active':1,'is_default':1})
    if bom_exists:
        bom_doc = frappe.get_doc('BOM', {'item':item,'is_active':1,'is_default':1})
    else:
        bom_doc=0
    return bom_doc

def bom_items(bom_doc):
    items=[]
    if bom_doc is not 0:
        for row in bom_doc.items:
            bom_doc=get_bom(row.item_code)
            if bom_doc is not 0: 
                items.append({
                    'item_code':row.item_code,
                    'qty':row.qty,
                    'bom_no':bom_doc.name,
                    'stock_uom':row.stock_uom,
                    'description':row.description,
                })
    return items

def bom_child_items(bom_doc,qty=None):
    items=[]
    if bom_doc is not 0:
        for row in bom_doc.items:
            bom_doc=get_bom(row.item_code)
            if bom_doc is not 0: 
                items.append({
                    'item_code':row.item_code,
                    'qty':row.qty*qty,
                    'bom_no':bom_doc.name,
                    'stock_uom':row.stock_uom,
                    'description':row.description,
                })
    return items

def check_items_bom(items):
    itembomlist=[]
    for row in items:
        bom_exists=frappe.db.exists('BOM', {'item':row['item_code'],'is_active':1,'is_default':1})
        if bom_exists:
            bom_doc = frappe.get_doc('BOM', {'item':row['item_code'],'is_active':1,'is_default':1})
            itembomlist.append({
                    'item_code':row['item_code'],
                    'qty':row['qty'],
                    'bom_no':bom_doc.name,
                    'stock_uom':row['stock_uom'],
                    'description':row['description'],
                })
    return itembomlist

def get_all_items(doc):
    total_items=[]
    check=False
    bom_doc=get_bom(doc.production_item)
    if bom_doc is not 0:
        items=bom_items(bom_doc)
        for row in items:
            total_items.append(row)
        check=True
    while check == True:
        ttl_items=[]
        for row in items:
            if items:
                bom_do=get_bom(row['item_code'])
                if bom_do is not 0:
                    item=bom_child_items(bom_do,row['qty'])
                    items_bom_list=check_items_bom(item)
                    if len(items_bom_list)>0:
                        for itm in items_bom_list:
                            ttl_items.append(itm)
        items=[]
        for row in ttl_items:
            items.append(row)
            total_items.append(row)
        if len(ttl_items) is 0:
            check=False
    
    return total_items

File: sultex/public/js/work_order.js
frappe.ui.form.on('Work Order', {
    refresh: function (frm) {
        if (frm.doc.docstatus === 0) {
            set_custom_buttons(frm);
        }
    },
    sales_order:function(frm,dt,dn){
        if(frm.doc.sales_order){
            frappe.model.set_value('Work Order',frm.doc.name,'reference_work_order',frm.doc.sales_order)
        }
    },
    bom_no: function (frm) {
        setTimeout(() => {
            for (let row of frm.doc.required_items) {
                frappe.call({
                    async: false,
                    method: 'sultex.events.work_order.get_req_items_details',
                    args: {
                        item_code: row.item_code,
                        bom_no: frm.doc.bom_no
                    },
                    callback: function (r) {
                        if (r.message) {
                            row.uom = r.message[0];
                            row.required_qty = r.message[1] * frm.doc.qty;
                        }
                    }
                });
            }
            frm.refresh_field('required_items')
        }, 500);
    },

    production_item: function (frm) {
        frm.doc.work_order_secondary_uom = [];
        frm.refresh_field('work_order_secondary_uom')
        setTimeout(() => {
            set_secondary_uom(frm)
        }, 500);
    },

    qty: function (frm) {
        setTimeout(() => {
            set_secondary_qty(frm)
        }, 500);
    },
    planned_start_date:function(frm){
        if(frm.doc.planned_start_date){
            show_msg(frm)
        }
    },
    expected_delivery_date:function(frm){
        if(frm.doc.expected_delivery_date){
            show_msg(frm)
        }
    },
})

const show_msg=(frm)=>{
    if(frm.doc.planned_start_date && frm.doc.expected_delivery_date){
        if(frm.doc.planned_start_date > frm.doc.expected_delivery_date){
            frappe.call({
                async: false,
                method: 'sultex.events.work_order.stop_execution',
                args: {
                    planned_start_date : frm.doc.planned_start_date,
                    expected_delivery_date :frm.doc.expected_delivery_date,
                    name : frm.doc.name
                },
                callback: function (r) {
                    //on success 
                }
            });
        }
    }
};

const set_secondary_uom = (frm) => {
    const item_code = frm.doc.production_item;
    if (!item_code) return;


    frappe.call({
        method: 'sultex.events.work_order.get_secondary_uom',
        args: { item_code },
        callback: function (r) {
            const data = r.message;
            if (data) {
                for (const row of data) {
                    let child_row = frm.add_child("work_order_secondary_uom")
                    let cf = parseFloat(row.conversion_factor);
                    let qty = frm.doc.qty ? parseFloat(frm.doc.qty) : 0;
                    let secondary_qty = cf * qty;

                    child_row.uom = row.uom
                    child_row.conversion_factor = row.conversion_factor
                    child_row.qty = secondary_qty
                }

                frm.refresh_field('work_order_secondary_uom')
            }
        }
    });
}

const set_secondary_qty = (frm) => {
    if (!frm.doc.work_order_secondary_uom) return;

    for (const row of frm.doc.work_order_secondary_uom) {
        let cf = parseFloat(row.conversion_factor);
        let qty = frm.doc.qty ? parseFloat(frm.doc.qty) : 0;
        let secondary_qty = cf * qty;

        row.qty = secondary_qty
    }

    frm.refresh_field('work_order_secondary_uom')
}

const set_custom_buttons = function (frm) {
    frm.add_custom_button('Sales Order', () => {
        if (frm.doc.sales_order) {
            frm.data = []
            let dialog = new frappe.ui.Dialog({
                title: __("Sales Order Item"),
                fields: [
                    {
                        fieldname: "items", read_only: 1, label: 'Sales Order Item', fieldtype: "Table", cannot_add_rows: true, data: frm.data,
                        get_data: () => {
                            return frm.data
                        },
                        fields: [
                            {
                                fieldtype: 'Link',
                                fieldname: "item_code",
                                options: 'Item',
                                in_list_view: 1,
                                read_only: 1,
                                label: __('Item Code'),
                                columns: 2
                            },
                            {
                                fieldtype: 'Data',
                                fieldname: "item_name",
                                in_list_view: 1,
                                read_only: 1,
                                label: __('Item Name'),
                                columns: 2
                            },
                            {
                                fieldtype: 'Link',
                                fieldname: "item_group",
                                options:'Item Group',
                                in_list_view: 1,
                                read_only: 1,
                                label: __('Item Group'),
                                columns: 2
                            },
                            {
                                fieldtype: 'Link',
                                fieldname: "uom",
                                options: 'UOM',
                                in_list_view: 1,
                                read_only: 1,
                                label: __('UOM'),
                                columns: 1
                            },
                            {
                                fieldtype: 'Float',
                                fieldname: "qty",
                                read_only: 1,
                                in_list_view: 1,
                                columns: 1,
                                label: __('Qty')
                            },
                            {
                                fieldtype: 'Check',
                                fieldname: "check",
                                label: __('Select'),
                                in_list_view: 1,
                                columns: 1
                            },
                            {
                                fieldtype: 'Date',
                                fieldname: "delivery_date",
                                label: __('Delivery Date'),
                                in_list_view: 1,
                                columns: 1
                            }
                        ]
                    }
                ],
                primary_action_label: 'Get Items',
                primary_action(values) {
                    let checked = 0
                    let item = '';
                    let qty = '';
                    let d_date='';
                    let sales_order=frm.doc.sales_order;
                    
                    let work_order = '';
                    if (values.items) {
                        for (let row of values.items) {
                            if (row.check === 1) {
                                checked += 1;
                                item = row.item_code;
                                qty = row.qty;
                                work_order = row.work_order;
                                d_date=row.delivery_date
                            }
                        }
                        if (checked === 1) {
                            cur_frm.set_value('production_item', item);
                            cur_frm.set_value('qty', qty);
                            cur_frm.set_value('expected_delivery_date', d_date);
                            cur_frm.set_value('from_work_order', work_order);
                        } else if (checked > 1) {
                            frappe.msgprint('Please select only one item')
                        } else {
                            frappe.msgprint('Please select at least one item')
                        }
                    }
                    setTimeout(() => {
                        cur_frm.set_value("sales_order", sales_order) 
                    }, 2000);
                    cur_frm.refresh_fields('production_item', 'qty')
                    dialog.hide();
                }
            });

            frappe.call({
                async: false,
                method: "sultex.events.work_order.sales_order_query",
                args: {
                    sales_order : cur_frm.doc.sales_order,
                    work_order  :cur_frm.doc.name,
                    delivery_date   :cur_frm.doc.expected_delivery_date,
                },
                callback: function (r) {
                    if (r.message) {
                        for (let row of r.message) {
                            frappe.db.get_doc('Item', row.item_code)
                            .then(itm_doc => {
                                setTimeout(() => {
                                    dialog.fields_dict.items.df.data.push({
                                        "item_code": row.item_code,
                                        "item_name": row.item_name,
                                        "item_group": itm_doc.item_group,
                                        'uom':row.uom,
                                        "qty": row.qty,
                                        "delivery_date":row.delivery_date,
                                    });
                                    frm.data = dialog.fields_dict.items.df.data;
                                    dialog.fields_dict.items.grid.refresh();
                                }, 500);
                            });
                        }
                        dialog.show();
                        dialog.$wrapper.find('.modal-dialog').css("max-width", "80%");
                    }
                }
            });
        }
        else {
            frappe.msgprint('Please Select Sales Order First');
        }
    }, 'Get Item from');

    frm.add_custom_button('Work Order', () => {
        frm.data = []
        let dialog = new frappe.ui.Dialog({
            title: __("Work Order Item"),
            outerWidth: '100%',
            fields: [
                {
                    fieldname: "items", read_only: 1, label: 'Work Order Item', fieldtype: "Table", cannot_add_rows: true, data: frm.data,
                    get_data: () => {
                        return frm.data
                    },
                    fields: [
                        {
                            fieldtype: 'Link',
                            fieldname: "work_order",
                            options: 'Work Order',
                            in_list_view: 1,
                            read_only: 1,
                            label: __('Work Order'),
                            columns: 2
                        },
                        {
                            fieldtype: 'Link',
                            fieldname: "item_code",
                            options: 'Item',
                            in_list_view: 1,
                            read_only: 1,
                            label: __('Item Code'),
                            columns: 2
                        },
                        {
                            fieldtype: 'Data',
                            fieldname: "item_name",
                            in_list_view: 1,
                            read_only: 1,
                            label: __('Item Name'),
                            columns: 2
                        },
                        {
                            fieldtype: 'Link',
                            fieldname: "item_group",
                            options: 'Item Group',
                            in_list_view: 1,
                            read_only: 1,
                            label: __('Item Group'),
                            columns: 1
                        },
                        {
                            fieldtype: 'Float',
                            fieldname: "qty",
                            read_only: 1,
                            in_list_view: 1,
                            columns: 1,
                            label: __('Qty')
                        },
                        {
                            fieldtype: 'Check',
                            fieldname: "check",
                            label: __('Select'),
                            in_list_view: 1,
                            columns: 1
                        }]
                }
            ],
            primary_action_label: 'Get Items',
            primary_action(values) {
                let checked = 0
                let item = '';
                let qty = '';
                let uom='';
                let item_group='';
                let work_order = '';
                if (values.items) {
                    for (let row of values.items) {
                        if (row.check === 1) {
                            checked += 1;
                            item = row.item_code;
                            uom=row.uom;
                            item_group=row.item_group;
                            qty = row.qty;
                            work_order = row.work_order;
                        }
                    }
                    if (checked === 1) {
                        cur_frm.set_value('production_item', item);
                        cur_frm.set_value('qty', qty);
                        cur_frm.set_value('from_work_order', work_order);
                    } else if (checked > 1) {
                        frappe.msgprint('Please select only one item')
                    } else {
                        frappe.msgprint('Please select at least one item')
                    }
                }
                cur_frm.refresh_fields('production_item', 'qty', 'from_work_order')
                dialog.hide();
            }
        });
        frappe.call({
            async: false,
            method: "sultex.events.work_order.work_order_query",
            callback: function (r) {
                if (r.message) {
                    for (let row of r.message) {
                        frappe.db.get_doc('Item', row.item_code)
                        .then(itm_doc => {
                            if(!itm_doc.is_purchase_item){
                                dialog.fields_dict.items.df.data.push({
                                    "work_order": row.parent,
                                    "item_code": row.item_code,
                                    "item_name": row.item_name,
                                    "item_group": itm_doc.item_group,
                                    "qty": row.required_qty
                                });
                                frm.data = dialog.fields_dict.items.df.data;
                                dialog.fields_dict.items.grid.refresh();
                            }
                        })
                        
                    }
                    dialog.show();
                    dialog.$wrapper.find('.modal-dialog').css("width", "80%");
                }
            }
        });
    }, 'Get Item from');


    // frm.add_custom_button('Work Order', () => {
    //     new frappe.ui.form.MultiSelectDialog({
    //       doctype: "Work Order",
    //       target: cur_frm,
    //       setters: {
    //           company: frm.doc.company,
    //           qty: frm.doc.qty
    //       },
    //       date_field: "creation",
    //       get_query() {
    //           return {
    //               filters: { docstatus: 1 }
    //           }
    //       },
    //       action(selections) {
    //           if (selections.length === 1) {
    //               frappe.call('sultex.events.work_order.fetch_items_from_work_order', {'work_order': selections})
    //               .then( r => {
    //                   if (r.message) {
    //                       frm.set_value('production_item', r.message.item);
    //                       frm.set_value('qty', r.message.qty);
    //                   }
    //                   frm.refresh_fields('production_item', 'qty')
    //               })
    //           } else {
    //               frappe.msgprint('Must Select one Work Order')
    //           }
    //           cur_dialog.hide();
    //       }
    //   });
    // }, 'Get Item from')
}